# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [155]:
# !pip install sqlalchemy

In [156]:
# Import your libraries:

%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sqlalchemy import create_engine, text


# Challenge 0 - Load,Query and Create connection of your Dataset

#### A lot of the times you won't have files already saved in an Excel or CSV file for you to prepare your data, implying that a lot of times you'll be extracting data from a Data Warehouse or Data Lake, majority of times through SQL.
#### A couple of times you may want to do some queries on your table in mySQL to get a slight view on what you have in hands,
#### so let's simulate that!

#### First we'll need to create a database and table in mySQL:

##### 1º- Open the austin_weather.sql file in MySQL Workbench and run the script into a desired schema.

#### 2º- As we are in mySQL Workbench, we can do some queries there, to get an overview on some characteristics of our data:
 - a) How many days are recorded in the dataset?
 - b) What is the day with the Highest Temperature in Fahrenheit (column TempHighF)
 - c) What is the average Humidity across all days? (column HumidityAvgPercent)
 - d) Top 10 days, where SeaLevelPressureAvgInches is the highest, knowing DewPointAvgF is higher than 28 ?

In [157]:
# Write your answers below:

# a) select count(distinct date(date)) from weather_data.austin_weather; #1319
# b)select date FROM weather_data.austin_weather
# where TempHighF = (
# select max(TempHighF) FROM weather_data.austin_weather); # 2017-07-29 00:00:00
# c)Select avg(HumidityAvgPercent) FROM weather_data.austin_weather; #66.57088703563305
# d)SELECT date 
# FROM weather_data.austin_weather
# WHERE DewPointAvgF > 28
# ORDER BY SeaLevelPressureAvgInches desc
# LIMIT 10;

#### Now that you explored a couple of elements in your table, let's bring your table into this jupyter notebook, by creating a Python-SQL connection like you did on MySQL Project!
#### In case you need a little refresher check this [link](https://www.dataquest.io/blog/sql-insert-tutorial/).
##### 1º - Create a connection using sqlalchemy from python to mysql 
##### 2º- Load the table into a variable called weather_df


In [158]:
# Your code here
import getpass

# password = getpass.getpass()

engine = create_engine("mysql+pymysql://root:user1234@localhost/weather_data",
        pool_recycle=3600,
        echo=True,)

In [159]:
connection = engine.connect()


2025-05-25 14:58:33,768 INFO sqlalchemy.engine.Engine SELECT DATABASE()
2025-05-25 14:58:33,769 INFO sqlalchemy.engine.Engine [raw sql] {}
2025-05-25 14:58:33,771 INFO sqlalchemy.engine.Engine SELECT @@sql_mode
2025-05-25 14:58:33,772 INFO sqlalchemy.engine.Engine [raw sql] {}
2025-05-25 14:58:33,773 INFO sqlalchemy.engine.Engine SELECT @@lower_case_table_names
2025-05-25 14:58:33,774 INFO sqlalchemy.engine.Engine [raw sql] {}


In [160]:
q ="SELECT * FROM austin_weather"
weather_df = pd.DataFrame(connection.execute(text(q)))

2025-05-25 14:58:33,785 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2025-05-25 14:58:33,787 INFO sqlalchemy.engine.Engine SELECT * FROM austin_weather
2025-05-25 14:58:33,787 INFO sqlalchemy.engine.Engine [generated in 0.00227s] {}


In [161]:
weather_df.sample(5)
weather_df.shape

(1319, 21)


# Challenge 1 - Describe the Dataset

#### Describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [162]:
# Your code here
print(weather_df.describe(), weather_df.dtypes)
print(weather_df.columns)

                      Date    TempHighF     TempAvgF     TempLowF
count                 1319  1319.000000  1319.000000  1319.000000
mean   2015-10-11 00:00:00    80.862775    70.642911    59.902957
min    2013-12-21 00:00:00    32.000000    29.000000    19.000000
25%    2014-11-15 12:00:00    72.000000    62.000000    49.000000
50%    2015-10-11 00:00:00    83.000000    73.000000    63.000000
75%    2016-09-04 12:00:00    92.000000    83.000000    73.000000
max    2017-07-31 00:00:00   107.000000    93.000000    81.000000
std                    NaN    14.766523    14.045904    14.190648 Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                         object
DewPointAvgF                          object
DewPointLowF                          object
HumidityHighPercent                   object
HumidityAvgPercent                    object
H

In [163]:
# Your code here
print(weather_df.head(5))

        Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHigh

In [164]:
# Your code here

def change_float(df, col_list):
    for col in col_list:
        df[col] = pd.to_numeric(df[col], errors='coerce').fillna('NaN').astype(float)
    return df

obj_col = weather_df.select_dtypes(include=['object']).columns
obj_col = obj_col.drop('Events')
print(obj_col)

weather_df_int = change_float(weather_df, obj_col)

weather_df_int.describe()

Index(['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'PrecipitationSumInches'],
      dtype='object')


Date    TempHighF     TempAvgF     TempLowF  \
count                 1319  1319.000000  1319.000000  1319.000000   
mean   2015-10-11 00:00:00    80.862775    70.642911    59.902957   
min    2013-12-21 00:00:00    32.000000    29.000000    19.000000   
25%    2014-11-15 12:00:00    72.000000    62.000000    49.000000   
50%    2015-10-11 00:00:00    83.000000    73.000000    63.000000   
75%    2016-09-04 12:00:00    92.000000    83.000000    73.000000   
max    2017-07-31 00:00:00   107.000000    93.000000    81.000000   
std                    NaN    14.766523    14.045904    14.190648   

       DewPointHighF  DewPointAvgF  DewPointLowF  HumidityHighPercent  \
count    1312.000000   1312.000000   1312.000000          1317.000000   
mean       61.515244     56.632622     50.933689            87.872437   
min        13.000000      8.000000      2.000000            37.000000   
25%        53.000000     46.000000     38.000000            85.000000   
50%        66.000000     61.000000     56.000000            90.000000   
75%        73.000000     69.000000     65.000000            94.000000   
max        80.000000     76.000000     75.000000           100.000000   
std        13.586452     14.866272     16.187480            11.038637   

       HumidityAvgPercent  HumidityLowPercent  SeaLevelPressureHighInches  \
count         1317.000000         1317.000000                 1316.000000   
mean            66.671982           44.958998                   30.112204   
min             27.000000           10.000000                   29.630000   
25%             59.000000           33.000000                   29.990000   
50%             67.000000           44.000000                   30.080000   
75%             74.000000           55.000000                   30.210000   
max             97.000000           93.000000                   30.830000   
std             12.463634           16.974632                    0.180149   

       SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
count                1316.000000                1316.000000   
mean                   30.022903                  29.931892   
min                    29.550000                  29.410000   
25%                    29.910000                  29.820000   
50%                    30.000000                  29.910000   
75%                    30.100000                  30.020000   
max                    30.740000                  30.610000   
std                     0.172145                   0.172144   

       VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
count          1307.000000         1307.000000         1307.000000   
mean              9.991584            9.162204            6.840092   
min               5.000000            2.000000            0.000000   
25%              10.000000            9.000000            3.000000   
50%              10.000000           10.000000            9.000000   
75%              10.000000           10.000000           10.000000   
max              10.000000           10.000000           10.000000   
std               0.163489            1.458883            3.683453   

       WindHighMPH   WindAvgMPH  WindGustMPH  PrecipitationSumInches  
count  1317.000000  1317.000000  1315.000000             1195.000000  
mean     13.231587     5.008352    21.373384                0.126268  
min       6.000000     1.000000     9.000000                0.000000  
25%      10.000000     3.000000    17.000000                0.000000  
50%      13.000000     5.000000    21.000000                0.000000  
75%      15.000000     6.000000    25.000000                0.010000  
max      29.000000    12.000000    57.000000                5.200000  
std       3.425096     2.086450     5.875657                0.448189

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [165]:
# Your code here
print(weather_df['DewPointAvgF'].unique())
weather_df['DewPointAvgF'].apply(type).value_counts()
weather_df[weather_df['DewPointAvgF'].isna()]


[49. 36. 27. 28. 40. 39. 41. 26. 42. 22. 48. 32.  8. 11. 45. 55. 61. 37.
 47. 25. 23. 20. 33. 30. 29. 17. 14. 13. 54. 59. 15. 24. 34. 35. 57. 50.
 53. 60. 46. 56. 51. 31. 38. 62. 43. 63. 64. 67. 66. 58. 70. 68. 65. 69.
 71. 72. nan 73. 74. 21. 44. 52. 12. 75. 76. 18.]


Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
174 2014-06-13         89        79        68            NaN           NaN   
175 2014-06-14         87        92        73            NaN           NaN   
176 2014-06-15         91        83        74            NaN           NaN   
177 2014-06-16         92        84        75            NaN           NaN   
596 2015-08-09        103        89        74            NaN           NaN   
597 2015-08-10        105        90        74            NaN           NaN   
598 2015-08-11        105        90        75            NaN           NaN   

     DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
174           NaN                 95.0                75.0   
175           NaN                 95.0                77.0   
176           NaN                  NaN                 NaN   
177           NaN                 94.0                72.0   
596           NaN                100.0                65.0   
597           NaN                100.0                62.0   
598           NaN                  NaN                 NaN   

     HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
174                55.0  ...                      29.95   
175                59.0  ...                      29.93   
176                 NaN  ...                      29.90   
177                49.0  ...                      29.99   
596                29.0  ...                        NaN   
597                24.0  ...                        NaN   
598                 NaN  ...                        NaN   

     SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
174                      29.87                  NaN                 NaN   
175                      29.84                  NaN                 NaN   
176                      29.87                  NaN                 NaN   
177                      29.93                  NaN                 NaN   
596                        NaN                  NaN                 NaN   
597                        NaN                  NaN                 NaN   
598                        NaN                  NaN                 NaN   

     VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
174                 NaN         15.0         4.0         22.0   
175                 NaN         15.0         6.0         22.0   
176                 NaN         14.0         9.0         23.0   
177                 NaN         12.0         8.0         24.0   
596                 NaN          NaN         NaN          NaN   
597                 NaN         14.0         6.0         20.0   
598                 NaN          NaN         NaN          NaN   

     PrecipitationSumInches  Events  
174                     0.0          
175                     0.0          
176                     0.0          
177                     NaN          
596                     0.0          
597                     0.0          
598                     0.0          

[7 rows x 21 columns]

In [166]:
# Your observation here
# Probably presence of empty strings or missing values

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [167]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [168]:
# Your code here
def change_num(df, col_list):
    for col in col_list:
        df[col] = pd.to_numeric(df[col], errors='coerce')
    return df

austin = change_num(weather_df, wrong_type_columns)

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [169]:
# Your code here
austin.dtypes

Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                        float64
DewPointAvgF                         float64
DewPointLowF                         float64
HumidityHighPercent                  float64
HumidityAvgPercent                   float64
HumidityLowPercent                   float64
SeaLevelPressureHighInches           float64
SeaLevelPressureAvgInches            float64
SeaLevelPressureLowInches            float64
VisibilityHighMiles                  float64
VisibilityAvgMiles                   float64
VisibilityLowMiles                   float64
WindHighMPH                          float64
WindAvgMPH                           float64
WindGustMPH                          float64
PrecipitationSumInches               float64
Events                                object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [170]:
# Your code here
missing_data = austin[austin.isnull().any(axis=1)]

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [171]:
# Your code here
print("no. of rows in full and missing data are: ",austin.shape[0], missing_data.shape[0])

no. of rows in full and missing data are:  1319 136


#### Calculate the ratio of missing rows to total rows

In [172]:
# Your code here
missing_data.shape[0]/austin.shape[0]*100

10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [173]:
# Your code here
missing_data.isna().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [174]:
# Your code here
missing_data['PrecipitationSumInches'].isna().sum()/austin['PrecipitationSumInches'].shape[0]*100
# missing_data['PrecipitationSumInches'].shape


np.float64(9.401061410159212)

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [181]:
# Your code here 
austin.drop(['PrecipitationSumInches'], inplace=True, axis=1)

# Print `austin` to confirm the column is indeed removed

austin

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    2013-12-21         74        60        45           67.0          49.0   
1    2013-12-22         56        48        39           43.0          36.0   
2    2013-12-23         58        45        32           31.0          27.0   
3    2013-12-24         61        46        31           36.0          28.0   
4    2013-12-25         58        50        41           44.0          40.0   
...         ...        ...       ...       ...            ...           ...   
1314 2017-07-27        103        89        75           71.0          67.0   
1315 2017-07-28        105        91        76           71.0          64.0   
1316 2017-07-29        107        92        77           72.0          64.0   
1317 2017-07-30        106        93        79           70.0          68.0   
1318 2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               E

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [182]:
# Your code here
austin_fixed = austin.interpolate(inplace=False)

/var/folders/lw/14dpn8b91z99svtdtfqnx09800tlp4/T/ipykernel_1617/2951143596.py:2: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  austin_fixed = austin.interpolate(inplace=False)


#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [191]:
# Your code here
austin.isna().any(axis=0)
austin_fixed.isna().sum()
austin_fixed.sample(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
901  2016-06-09         92        82        71           72.0          70.0   
199  2014-07-08         98        86        74           73.0          68.0   
1079 2016-12-04         53        51        49           49.0          48.0   
214  2014-07-23         97        86        75           75.0          71.0   
771  2016-01-31         86        72        58           58.0          54.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
901           67.0                 93.0                73.0   
199           61.0                 93.0                63.0   
1079          46.0                 93.0                88.0   
214           66.0                 94.0                66.0   
771           36.0                 93.0                55.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
901                 52.0                       29.96   
199                 33.0                       30.04   
1079                83.0                       30.07   
214                 38.0                       30.05   
771                 17.0                       29.87   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
901                       29.92                      29.87   
199                       29.99                      29.95   
1079                      30.01                      29.93   
214                       30.00                      29.94   
771                       29.81                      29.73   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
901                  10.0                10.0                 6.0   
199                  10.0                10.0                10.0   
1079                 10.0                 8.0                 2.0   
214                  10.0                 9.0                 6.0   
771                  10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH Events  
901          20.0         5.0         34.0         
199          13.0         7.0         21.0         
1079         13.0         6.0         19.0   Rain  
214           7.0         3.0         14.0         
771          12.0         5.0         19.0

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [192]:
# Your code here:
austin_fixed['Events'].value_counts()

Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [ ]:
# Your answer:
 # 3

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [193]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for event in event_list:
    austin_fixed[event] = 0

# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    2013-12-21         74        60        45           67.0          49.0   
1    2013-12-22         56        48        39           43.0          36.0   
2    2013-12-23         58        45        32           31.0          27.0   
3    2013-12-24         61        46        31           36.0          28.0   
4    2013-12-25         58        50        41           44.0          40.0   
...         ...        ...       ...       ...            ...           ...   
1314 2017-07-27        103        89        75           71.0          67.0   
1315 2017-07-28        105        91        76           71.0          64.0   
1316 2017-07-29        107        92        77           72.0          64.0   
1317 2017-07-30        106        93        79           70.0          68.0   
1318 2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
1                   43.0  ...                10.0                 5.0   
2                   27.0  ...                10.0                10.0   
3                   22.0  ...                10.0                 7.0   
4                   56.0  ...                10.0                 7.0   
...                  ...  ...                 ...                 ...   
1314                25.0  ...                10.0                10.0   
1315                20.0  ...                10.0                10.0   
1316                19.0  ...                10.0                10.0   
1317                27.0  ...                10.0                10.0   
1318                22.0  ...                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               Events  Snow  Fog  \
0            20.0         4.0         31.0  Rain , Thunderstorm     0    0   
1            16.0         6.0         25.0                          0    0   
2             8.0         3.0         12.0                          0    0   
3            12.0         4.0         20.0                          0    0   
4            10.0         2.0         16.0                          0    0   
...           ...         ...          ...                  ...   ...  ...   
1314         12.0         5.0         21.0                          0    0   
1315         14.0         5.0         20.0                          0    0   
1316         12.0         4.0         17.0                          0    0   
1317         13.0         4.0         20.0                          0    0   
1318         12.0         4.0         20.0                          0    0   

      Rain Thunderstorm  
0        0            0  
1        0            0  
2        0            0  
3        0            0  
4        0            0  
...    ...          ...  
1314     0            0  
1315     0            0  
1316     0            0  
1317     0            0  
1318     0            0  

[1319 rows x 24 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [196]:
# Your code here
austin_fixed['Events'].value_counts()
austin_fixed['Events'].str.contains('Rain')

for event in event_list:
    austin_fixed[event] = austin_fixed['Events'].str.contains(event).astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [199]:
# Your code here
austin_fixed.sample(10)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
1097 2016-12-22         68        59        50           40.0          36.0   
812  2016-03-12         74        66        57           58.0          56.0   
1150 2017-02-13         78        71        63           64.0          55.0   
752  2016-01-12         66        50        34           39.0          34.0   
1013 2016-09-29         86        75        63           62.0          55.0   
909  2016-06-17         97        87        76           75.0          73.0   
661  2015-10-13         94        79        64           73.0          45.0   
1015 2016-10-01         80        70        59           61.0          56.0   
1060 2016-11-15         86        73        59           58.0          55.0   
352  2014-12-08         67        58        49           49.0          46.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
1097          33.0                 63.0                48.0   
812           54.0                100.0                75.0   
1150          42.0                 84.0                63.0   
752           29.0                 89.0                58.0   
1013          50.0                 87.0                61.0   
909           71.0                 94.0                70.0   
661           28.0                 79.0                45.0   
1015          53.0                 87.0                65.0   
1060          51.0                 87.0                60.0   
352           45.0                 86.0                69.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
1097                32.0  ...                10.0                 9.0   
812                 49.0  ...                10.0                 8.0   
1150                42.0  ...                10.0                10.0   
752                 27.0  ...                10.0                10.0   
1013                34.0  ...                10.0                10.0   
909                 46.0  ...                10.0                 9.0   
661                 10.0  ...                10.0                10.0   
1015                42.0  ...                10.0                10.0   
1060                32.0  ...                10.0                10.0   
352                 52.0  ...                10.0                 8.0   

      WindHighMPH  WindAvgMPH  WindGustMPH  Events  Snow  Fog  Rain  \
1097         12.0         4.0         19.0             0    0     0   
812           9.0         4.0         16.0    Rain     0    0     1   
1150         18.0         9.0         28.0             0    0     0   
752          10.0         2.0         16.0             0    0     0   
1013         12.0         5.0         21.0             0    0     0   
909          10.0         6.0         19.0             0    0     0   
661          12.0         5.0         20.0             0    0     0   
1015          7.0         1.0         13.0             0    0     0   
1060          8.0         2.0         14.0             0    0     0   
352           9.0         3.0         14.0             0    0     0   

     Thunderstorm  
1097            0  
812             0  
1150            0  
752             0  
1013            0  
909             0  
661             0  
1015            0  
1060            0  
352             0  

[10 rows x 24 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [ ]:
# Your code here
austin_fixed.drop(['Events'], inplace=True)
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    2013-12-21         74        60        45           67.0          49.0   
1    2013-12-22         56        48        39           43.0          36.0   
2    2013-12-23         58        45        32           31.0          27.0   
3    2013-12-24         61        46        31           36.0          28.0   
4    2013-12-25         58        50        41           44.0          40.0   
...         ...        ...       ...       ...            ...           ...   
1314 2017-07-27        103        89        75           71.0          67.0   
1315 2017-07-28        105        91        76           71.0          64.0   
1316 2017-07-29        107        92        77           72.0          64.0   
1317 2017-07-30        106        93        79           70.0          68.0   
1318 2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   57.0  ...                 10.0                 7.0   
1                   43.0  ...                 10.0                10.0   
2                   27.0  ...                 10.0                10.0   
3                   22.0  ...                 10.0                10.0   
4                   56.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1314                25.0  ...                 10.0                10.0   
1315                20.0  ...                 10.0                10.0   
1316                19.0  ...                 10.0                10.0   
1317                27.0  ...                 10.0                10.0   
1318                22.0  ...                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1314                10.0         12.0         5.0         21.0     0    0   
1315                10.0         14.0         5.0         20.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1317                10.0         13.0         4.0         20.0     0    0   
1318                10.0         12.0         4.0         20.0     0    0   

      Rain  Thunderstorm  
0        1             1  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1314     0             0  
1315     0             0  
1316     0             0  
1317     0             0  
1318     0             0  

[1319 rows x 23 columns]

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [233]:
# Your code here
from datetime import datetime
austin_fixed.dtypes
austin_fixed['Date']
datetime.strptime('2013-12-21','%Y-%m-%d').date().toordinal()
# austin_fixed['Date'] = pd.to_datetime(austin_fixed['Date'],'%Y-%m-%d').apply(lambda x: x.toordinal())
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     735223         74        60        45           67.0          49.0   
1     735224         56        48        39           43.0          36.0   
2     735225         58        45        32           31.0          27.0   
3     735226         61        46        31           36.0          28.0   
4     735227         58        50        41           44.0          40.0   
...      ...        ...       ...       ...            ...           ...   
1314  736537        103        89        75           71.0          67.0   
1315  736538        105        91        76           71.0          64.0   
1316  736539        107        92        77           72.0          64.0   
1317  736540        106        93        79           70.0          68.0   
1318  736541         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityHighMiles  VisibilityAvgMiles  \
0                   57.0  ...                 10.0                 7.0   
1                   43.0  ...                 10.0                10.0   
2                   27.0  ...                 10.0                10.0   
3                   22.0  ...                 10.0                10.0   
4                   56.0  ...                 10.0                10.0   
...                  ...  ...                  ...                 ...   
1314                25.0  ...                 10.0                10.0   
1315                20.0  ...                 10.0                10.0   
1316                19.0  ...                 10.0                10.0   
1317                27.0  ...                 10.0                10.0   
1318                22.0  ...                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                    2.0         20.0         4.0         31.0     0    0   
1                    5.0         16.0         6.0         25.0     0    0   
2                   10.0          8.0         3.0         12.0     0    0   
3                    7.0         12.0         4.0         20.0     0    0   
4                    7.0         10.0         2.0         16.0     0    0   
...                  ...          ...         ...          ...   ...  ...   
1314                10.0         12.0         5.0         21.0     0    0   
1315                10.0         14.0         5.0         20.0     0    0   
1316                10.0         12.0         4.0         17.0     0    0   
1317                10.0         13.0         4.0         20.0     0    0   
1318                10.0         12.0         4.0         20.0     0    0   

      Rain  Thunderstorm  
0        1             1  
1        0             0  
2        0             0  
3        0             0  
4        0             0  
...    ...           ...  
1314     0             0  
1315     0             0  
1316     0             0  
1317     0             0  
1318     0             0  

[1319 rows x 23 columns]

#### Print `austin_fixed` to check your `Date` column.

In [232]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0  ...                 10.0                 7.0                 2.0   
1  ...                 10.0                10.0                 5.0   
2  ...                 10.0                10.0                10.0   
3  ...                 10.0                10.0                 7.0   
4  ...                 10.0                10.0                 7.0   

   WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
0         20.0         4.0         31.0     0    0     1             1  
1         16.0         6.0         25.0     0    0     0             0  
2          8.0         3.0         12.0     0    0     0             0  
3         12.0         4.0         20.0     0    0     0             0  
4         10.0         2.0         16.0     0    0     0             0  

[5 rows x 23 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [238]:
# Your code here:
X = austin_fixed.drop(['TempAvgF', 'Date'], axis=1)
X.columns

y = austin_fixed[['TempAvgF']]
y.columns

Index(['TempAvgF'], dtype='object')

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [240]:
#Your code here:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.20, random_state=1)

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [243]:
#Your code here:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(1055, 21) (264, 21) (1055, 1) (264, 1)


#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [ ]:
# Your code here:


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [ ]:
# Your code here:
